### PR TITLE
Optimize snapshot storage

### PR DESCRIPTION
**Pull request by viliam-durina**

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/StoreSnapshotTasklet.java [hunk 1]
```diff
@@ -79,64 +83,75 @@ private void stateMachineStep() {
...
+                    progTracker.notDone();
+                } else {
+                    progTracker.madeProgress();
                     snapshotContext.snapshotDoneForTasklet();
                     pendingSnapshotId++;
-                    mapWriter.setMapName(currMapName());
-                    state = inputIsDone ? DONE : DRAIN;
+                    resetCurrentMap();
                     hasReachedBarrier = false;
+                    if (inputIsDone) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/StoreSnapshotTasklet.java]: I think it was better with two explicit states. The code is hard to follow.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/StoreSnapshotTasklet.java [hunk 2]
```diff
@@ -79,64 +83,75 @@ private void stateMachineStep() {
...
-                        mapWriter.put((Entry<Data, Data>) o);
+                        if (!ssWriter.offer((Entry<Data, Data>) o)) {
+                            pendingEntry = (Entry<Data, Data>) o;
+                            return false;
+                        }
                     }
+                    return true;
                 });
                 if (result.isDone()) {
+                    state = REACHED_BARRIER;
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/StoreSnapshotTasklet.java]: see the other comment about explicit states.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Map.Entry;
+
+public interface AsyncSnapshotWriter {
+    void setCurrentMap(String mapName);
+
+    @CheckReturnValue
+    boolean offer(Entry<? extends Data, ? extends Data> entry);
+
+    @CheckReturnValue
+    boolean flushRemaining();
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java]: I would simply prefer the name `flush` , remaining is redundant.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * @throws Throwable Any error occurred during writing to underlying map.
+     * Error is reported only once, next call will say whether the pending flushes are done.
+     */
+    boolean hasPendingFlushes() throws Throwable;
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java]: maybe the name `isReady()` is better. Error checking also could be separated into another method such as `hasErrors()`.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    boolean isEmpty();
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java]: `hasPendingEntries` perhaps?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    @Override
+    protected void init(@Nonnull Context context) {
+        serializationService =
+                ((HazelcastInstanceImpl) context.jetInstance().getHazelcastInstance()).getSerializationService();
+    }
+
+    private Traverser<Object> traverser(byte[] data) {
+        BufferObjectDataInput in;
+        try {
+            in = serializationService.createObjectDataInput(IOUtil.decompress(data));
```
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImpl.java [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+        @Override
+        public void readData(ObjectDataInput in) {
+        }
+    }
+
+    /**
+     * Variant of {@code java.io.ByteArrayOutputStream} with capacity limit.
+     */
+    static class CustomByteArrayOutputStream extends OutputStream {
```
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImpl.java]: It's a hard limit in this class, see `ensureCapacity`. In other places I make sure not to exceed the limit, but this class checks it.

With `ByteArrayOutputStream`, you have no access to internal buffer. If I was appending to `ByteArrayOutputStream`, just in order to read it to compress it I would have to copy it: one more copying.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package com.hazelcast.jet.impl.util;
+
+import com.hazelcast.nio.serialization.Data;
+
+import javax.annotation.CheckReturnValue;
+import java.util.Map.Entry;
+
+public interface AsyncSnapshotWriter {
```
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java]: There's `MockAsyncSnapshotWriter` used in `StoreSnapshotTaskletTest`. This way I can test two complex classes independently. `StoreSnapshotTaskletTest` even doesn't need to start `JetInstsance`.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java [hunk 9]
```diff
@@ -174,12 +180,34 @@ public void memberAdded(MembershipServiceEvent event) {
...
+        return numConcurrentAsyncOps;
     }
 
     @Override
     public void populate(LiveOperations liveOperations) {
         liveOperationRegistry.populate(liveOperations);
     }
+
+    public int[] getSharedPartitionKeys() {
+        if (sharedPartitionKeys == null) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java]: can use `Util.memoizeConcurrent` instead

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java (hunk 4 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java (hunk 5 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriter.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java (hunk 9 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/StoreSnapshotTasklet.java (hunk 1 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/StoreSnapshotTasklet.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java [hunk 10]
```diff
@@ -63,7 +68,9 @@
     private JobCoordinationService jobCoordinationService;
     private JobExecutionService jobExecutionService;
 
-    private final AtomicInteger numConcurrentPutAllOps = new AtomicInteger();
+    private final AtomicInteger numConcurrentAsyncOps = new AtomicInteger();
+
+    private volatile Supplier<int[]> sharedPartitionKeys = Util.memoizeConcurrent(this::computeSharedPartitionKeys);
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java]: should be `final` instead of `volatile`

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+                ((HazelcastInstanceImpl) context.jetInstance().getHazelcastInstance()).getSerializationService();
+    }
+
+    private Traverser<Object> traverser(byte[] data) {
+        BufferObjectDataInput in;
+        in = serializationService.createObjectDataInput(data);
+
+        return () -> uncheckCall(() -> {
+            Object key = in.readObject();
+            if (key == SnapshotDataValueTerminator.INSTANCE) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java]: should close the buffer here.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        serializationService =
+                ((HazelcastInstanceImpl) context.jetInstance().getHazelcastInstance()).getSerializationService();
+    }
+
+    private Traverser<Object> traverser(byte[] data) {
+        BufferObjectDataInput in;
+        in = serializationService.createObjectDataInput(data);
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java]: these two lines could be merged

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java [hunk 13]
```diff
@@ -230,26 +226,19 @@ void tryStartJob(Function<Long, Long> executionIdSupplier) {
...
             // Processor.finishSnapshotRestore() method is always called on all vertices in
             // a job which is restored from a snapshot.
             String mapName = snapshotDataMapName(jobId, snapshotId, vertex.getName());
             Vertex readSnapshotVertex = dag.newVertex("__snapshot_read." + vertex.getName(), readMapP(mapName));
-            // We need a separate mapping vertex and can't use readMapP's projectionFn:
-            // the projection will cause key/value deserialization on partition thread, which doesn't have job's
-            // class loader. If the key/value uses a custom object, it will fail. For example, StreamKafkaP uses
-            // TopicPartition as the key or a custom processor can use custom key.
-            Vertex mapSnapshotVertex = dag.newVertex("__snapshot_map_." + vertex.getName(), mapP(projection));
+            Vertex explodeVertex = dag.newVertex("_snapshot_explode." + vertex.getName(), ExplodeSnapshotP::new);
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java]: should be `__snapshot_` to match the name of the read vertex

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java [hunk 14]
```diff
@@ -230,26 +226,19 @@ void tryStartJob(Function<Long, Long> executionIdSupplier) {
...
 
             readSnapshotVertex.localParallelism(vertex.getLocalParallelism());
-            mapSnapshotVertex.localParallelism(vertex.getLocalParallelism());
+            explodeVertex.localParallelism(vertex.getLocalParallelism());
 
             int destOrdinal = dag.getInboundEdges(vertex.getName()).size();
-            dag.edge(between(readSnapshotVertex, mapSnapshotVertex).isolated())
-               .edge(new SnapshotRestoreEdge(mapSnapshotVertex, vertex, destOrdinal));
+            dag.edge(between(readSnapshotVertex, explodeVertex).isolated());
+            dag.edge(new SnapshotRestoreEdge(explodeVertex, vertex, destOrdinal));
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java]: could use the fluent API here as before

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java [hunk 15]
```diff
@@ -57,8 +59,6 @@
  */
 public class AsyncMapWriter {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java]: is this class still used anywhere?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java (hunk 11 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java (hunk 12 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ExplodeSnapshotP.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java (hunk 13 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java (hunk 14 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java]: fixed

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java (hunk 15 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java]: No, it's not. I kept it because we can use it for local imap sink.
